lib/JSON/Validator/Schema/OpenAPIv3.pm

```diff
@@ -0,0 +1,214 @@
+    'json-pointer'  => JSON::Validator::Formats->can('check_json_pointer'),
+    'password'      => sub {undef},
+    'regex'         => JSON::Validator::Formats->can('check_regex'),
+    'relative-json-pointer' =>
+      JSON::Validator::Formats->can('check_relative_json_pointer'),
+    'time'          => JSON::Validator::Formats->can('check_time'),
+    'uri'           => JSON::Validator::Formats->can('check_uri'),
+    'uri-reference' => JSON::Validator::Formats->can('check_uri_reference'),
+    'uri-template'  => JSON::Validator::Formats->can('check_uri_template'),
+  };
+}
+
+sub _build_request_parameter_rules {
+  my ($self, $method, $path) = @_;
+
+  my $cache_key = "$method:$path";
+  return $self->{request_parameters}{$cache_key}
+    if $self->{request_parameters}{$cache_key};
+
+  my @parameters
+    = map {@$_} $self->find_all_nodes([paths => $path, $method], 'parameters');
+
+  my $request_body = $self->get([paths => $path, $method, 'requestBody']);
+  if ($request_body) {
+    my $param = {in => 'body', name => 'body'};
+    $param->{accepts}  = $request_body->{content} || {};
+    $param->{required} = $request_body->{required};
+    push @parameters, $param;
+  }
+
+  return $self->{request_parameters}{$cache_key} = \@parameters;
+}
+
+sub _definitions_path_for_ref {
+  my ($self, $ref) = @_;
+
+  # Try to determine the path from the fqn
+  # We are only interested in the path in the fqn, so following fqn:
+  # "#/components/schemas/some_schema" => ['components', 'schemas']
+  my $path
+    = Mojo::Path->new($ref->fqn =~ m!^.*#/(components/.+)$!)->to_dir->parts;
+  return $path->[0] ? $path : ['components'];
+}
+
+sub _prefix_path {
+  return join '', "/$_[0]", $_[1] ? ($_[1]) : ();
+}
+
+sub _param_key {
+  return join ':', grep {defined} @{$_[0]}{qw(in name)};
+}
+
+sub _response_schema {
+  my ($self, $schema) = @_;
+  return {
+    description => 'Default response.',
+    content     => {'application/json' => {schema => $schema}},
+  };
+}
+
+sub _sub_schemas         { shift->data->{components}{schemas} ||= {} }
+sub _sub_schemas_pointer {'#/components/schemas'}
+
+sub _validate_request_parameters {
+  my ($self, $parameters, $request) = @_;
+  my @errors;
+
+  # Make it faster to lookup input values
+  $request = {map { (_param_key($_) => $_) } @$request};
+
+  # [in, name, exists, value, content_type] or [in, name, exists, value]
+  for my $param (@$parameters) {
+    my $req = $request->{_param_key($param)} || {};
+
+    if ($param->{required} and !$req->{exists}) {
+      push @errors, E "/$param->{name}", [qw(object required)];
+      next;
+    }
+    elsif ($req->{exists}) {
+      my $schema
+        = $param->{in} eq 'body'
+        ? $param->{accepts}{$req->{content_type}}{schema}
+        : $param->{schema};
+      push @errors,
+        map { $_->path(_prefix_path($param->{name}, $_->path)); $_ }
+        $self->validate($req->{value}, $schema);
+    }
+  }
+
+  return @errors;
+}
+
```

Comment:
Another missing part, again a bit of rewrite from previous code in M::P::OpenAPI.

This one is not complete, If Im not mistaken it has something to do with coerce. But I have not been able to figure it out entirely ;). A good unit test for this is `t/v3-basic.t` from M::P::OpenAPI

```perl
sub validate_response {
  my ($self, $c, $sub_schema, $status, $data) = @_;

  return E '/' => "No responses rules defined for status $status."
    unless my $res_schema
    = $sub_schema->{responses}{$status} || $sub_schema->{responses}{default};

  my $accept = $self->_negotiate_accept_header($c, $res_schema);
  return JSON::Validator::E('/' => "No responses rules defined for $accept.")
    unless $res_schema = $res_schema->{content}{$accept};
  $c->stash(openapi_negotiated_content_type => $accept);

  my @errors;
  push @errors, $self->_validate_response_headers($c, $res_schema->{headers})
    if $res_schema->{headers};

  if ($res_schema->{'x-json-schema'}) {
    warn "[OpenAPI] Validate using x-json-schema\n" if DEBUG;
    push @errors, $self->_validate($data, '', $res_schema->{'x-json-schema'});
  }
  elsif ($res_schema->{schema}) {
    warn "[OpenAPI] Validate using schema\n" if DEBUG;
    push @errors, $self->_validate($data, '', $res_schema->{schema});
  }

  return @errors;
}


sub _negotiate_accept_header {
  my ($self, $c, $schema) = @_;
  my $accept    = $c->req->headers->accept || '*/*';
  my @in_schema = sort { length $b <=> length $a } keys %{$schema->{content}};
  my (@from_req, %from_req);

  /^\s*([^,; ]+)(?:\s*\;\s*q\s*=\s*(\d+(?:\.\d+)?))?\s*$/i and $from_req{lc $1} = $2 // 1
    for split /,/, $accept;
  @from_req = sort { $from_req{$b} <=> $from_req{$a} } sort keys %from_req;

  # Check for exact match
  for my $ct (@from_req) {
    return $ct if $schema->{content}{$ct};
  }

  # Check for closest match
  for my $re (map { s!\*!.*!g; qr{$_} } grep {/\*/} @in_schema) {
    for my $ct (@from_req) {
      return $ct if $ct =~ $re;
    }
  }
  for my $re (map { s!\*!.*!g; qr{$_} } grep {/\*/} @from_req) {
    for my $ct (@in_schema) {
      return $ct if $ct =~ $re;
    }
  }

  # Could not find any valid content type
  return $accept;
}
```

Reply:
I'm not sure what you're referring to, since I force pushed over the change, but if it has to do with negotiating accept headers, then it must be done in M::P::OpenAPI.

Reply:
> I'm not sure what you're referring to, since I force pushed over the change, but if it has to do with negotiating accept headers, then it must be done in M::P::OpenAPI.

This was referring to an older version. But yes, this belongs in M::P::OpenAPI